#!/bin/bash -l
#PBS -N htseq_alm
#PBS -l walltime=10:00:00
#PBS -l mem=50G
#PBS -l ncpus=1
#PBS -W depend=afterok:
#PBS -J 1-69

cd d:\Libraries\Documents\GitHub\Various_scripts\DGE\htseq_count

## MANUAL SETUP BELOW
## SETUP: Specify mapped reads SAM files
SAMFILES="d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\1_18-12-2019_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\2_18-12-2019_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\3_18-12-2019_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\5_18-12-2019_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\6_18-12-2019_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\7_18-12-2019_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\9_02-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\10_02-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\11_02-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\13_02-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\14_02-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\15_02-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\17_16-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\18_16-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\19_16-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\20_16-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\21_16-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\22_16-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\23_16-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\24_16-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\25_30-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\26_30-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\27_30-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\28_30-01-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\29_30-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\30_30-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\31_30-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\32_30-01-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\33_13-02-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\34_13-02-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\35_13-02-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\37_13-02-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\38_13-02-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\39_13-02-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\41_12-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\42_12-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\43_12-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\44_12-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\45_12-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\46_12-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\47_12-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\48_12-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\49_26-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\50_26-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\51_26-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\52_26-03-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\53_26-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\54_26-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\55_26-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\56_26-03-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\57_09-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\58_09-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\59_09-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\60_09-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\61_09-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\62_09-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\63_09-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\64_09-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\65_23-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\66_23-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\67_23-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\68_23-04-2020_Leaf/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\69_23-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\70_23-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\71_23-04-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\73_07-05-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\74_07-05-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\75_07-05-2020_Bud/Aligned.out.sam d:\Libraries\Documents\GitHub\Various_scripts\DGE\star_map\76_07-05-2020_Bud/Aligned.out.sam"
ARRAY1=($SAMFILES)

## SETUP: Sample names (paired to SAMFILES)
SAMNAMES="1_18-12-2019_Leaf 2_18-12-2019_Leaf 3_18-12-2019_Leaf 5_18-12-2019_Bud 6_18-12-2019_Bud 7_18-12-2019_Bud 9_02-01-2020_Leaf 10_02-01-2020_Leaf 11_02-01-2020_Leaf 13_02-01-2020_Bud 14_02-01-2020_Bud 15_02-01-2020_Bud 17_16-01-2020_Leaf 18_16-01-2020_Leaf 19_16-01-2020_Leaf 20_16-01-2020_Leaf 21_16-01-2020_Bud 22_16-01-2020_Bud 23_16-01-2020_Bud 24_16-01-2020_Bud 25_30-01-2020_Leaf 26_30-01-2020_Leaf 27_30-01-2020_Leaf 28_30-01-2020_Leaf 29_30-01-2020_Bud 30_30-01-2020_Bud 31_30-01-2020_Bud 32_30-01-2020_Bud 33_13-02-2020_Leaf 34_13-02-2020_Leaf 35_13-02-2020_Leaf 37_13-02-2020_Bud 38_13-02-2020_Bud 39_13-02-2020_Bud 41_12-03-2020_Leaf 42_12-03-2020_Leaf 43_12-03-2020_Leaf 44_12-03-2020_Leaf 45_12-03-2020_Bud 46_12-03-2020_Bud 47_12-03-2020_Bud 48_12-03-2020_Bud 49_26-03-2020_Leaf 50_26-03-2020_Leaf 51_26-03-2020_Leaf 52_26-03-2020_Leaf 53_26-03-2020_Bud 54_26-03-2020_Bud 55_26-03-2020_Bud 56_26-03-2020_Bud 57_09-04-2020_Leaf 58_09-04-2020_Leaf 59_09-04-2020_Leaf 60_09-04-2020_Leaf 61_09-04-2020_Bud 62_09-04-2020_Bud 63_09-04-2020_Bud 64_09-04-2020_Bud 65_23-04-2020_Leaf 66_23-04-2020_Leaf 67_23-04-2020_Leaf 68_23-04-2020_Leaf 69_23-04-2020_Bud 70_23-04-2020_Bud 71_23-04-2020_Bud 73_07-05-2020_Bud 74_07-05-2020_Bud 75_07-05-2020_Bud 76_07-05-2020_Bud"
ARRAY2=($SAMNAMES)

# SETUP: Specify Python2 location
PY2DIR=D:\Bioinformatics\Anaconda_2

# SETUP: GFF annotation file
GFF=F:\plant_rnaseq\annotations\Prudul26A.chromosomes.gff3

# SETUP: Output file name

## MANUAL SETUP END

## RUN PROGRAM
ARRAY_INDEX=$((${PBS_ARRAY_INDEX}-1))
FILE=${ARRAY1[${ARRAY_INDEX}]}
NAME=${ARRAY2[${ARRAY_INDEX}]}
OUT=$NAME.htseq.counts

$PY2DIR/python2.7 -m HTSeq.scripts.count -r name -a 0 -t gene -i ID $FILE $GFF > $OUT
    